(#230) Add Admin Only Source For Chocolatey Core Packages

ryanrichter94 opened the pull request:
## Description Of Changes
- Add ChocolateyCore admin only source
- Add ChocolateyCore NuGet repo within Nexus
- Add Jenkins job Update ChocolateyCore Repository
- Change previous C:\choco-setup\files\files directory to be C:\choco-setup\files\packages

## Motivation and Context
Main reason for creating this repository was from feedback of admins not wanting their end users to see the Chocolatey core packages in the default repository. This way ChocolateyInternal only contains the packages brought in by the Jenkins internalization pipeline. Also cleaning up the end user view.

## Testing
1. Tested on local VM.

### Operating Systems Testing
- Windows Server 2022

## Change Types Made

* ~[ ] Bug fix (non-breaking change).~
* ~[ ] Feature / Enhancement (non-breaking change).~
* [X] Breaking change (fix or feature that could cause existing functionality to change).
* [X] Documentation changes.
* [X] PowerShell code changes.

## Change Checklist

* [x] Requires a change to the documentation.
* [X] Documentation has been updated.
* [x] Tests to cover my changes, have been added.
* [x] All new and existing tests passed?
* [x] PowerShell code changes: PowerShell v3 compatibility checked?

## Related Issue
- Docs PR with README changes: https://github.com/chocolatey/docs/pull/1038

Fixes #230

ryanrichter94 committed:
(#230) Adds ChocolateyCore Repository and Jenkins Job

- Add ChocolateyCore repo as main repo for core packages
- Add callout to new ChocolateyCore repository
- Remove See It In Action README section as video is outdated
- Fixup spelling/grammar in README
- Add new Update ChocolateyCore Repository Jenkins Job
- Add test to check for new Jenkins job
- Remove legacy build files from new and existing Jenkins jobs

ryanrichter94 committed:
(fix) Fix Flaky Edge Package Test

Add handling to bring edge under Chocolatey management if it was natively installed by a newer server OS.

This should ensure that the test is more consistent, delivering the same result regardless of if the Chocolatey package were installed or not.

ryanrichter94 commented:
Converted back to draft after finding issue when running ClientSetup script.

JPRuskin commented on scripts/ClientSetup.ps1:
```diff
@@ -111,6 +111,13 @@ $script = $webClient.DownloadString("https://${hostAddress}/repository/choco-ins
+}
+
 if ($InternetEnabled) {
     choco source add --name="'ChocolateyInternal'" --source="'$RepositoryUrl'" --allow-self-service --user="'$($Credential.UserName)'" --password="'$($Credential.GetNetworkCredential().Password)'" --priority=1
```
So, this is adding the second source with the same URL as the block above with `ChocolateyCore` unless I'm mistaken?

Do we need to pass in both URLs?

JPRuskin replied:
Also, we were turning off self-service/making the source admin only for the Internal feed, right?

ryanrichter94 replied:
Sorted with Regex

JPRuskin commented on Set-SslSecurity.ps1:
```diff
@@ -271,7 +297,7 @@ process {
 # Touch NOTHING below this line
 `$User = 'chocouser'
 `$SecurePassword = `$NexusUserPW | ConvertTo-SecureString -AsPlainText -Force
-`$RepositoryUrl = "https://`$(`$fqdn):8443/repository/ChocolateyInternal/index.json"
+`$RepositoryUrl = "https://`$(`$fqdn):8443/repository/ChocolateyCore/index.json"
```
Do we need to keep _a_ definition of the Internal repo? We're still using it as an admin only source, right?

ryanrichter94 replied:
Need both as one is base artifacts and other is default output from Jenkins jobs.

JPRuskin reviewed (commented):
Sorry, delayed hitting send on this while you were out but remembered.

I think we need to change up how we're adding the internal source a little.

JPRuskin committed:
(#230) Change "files" directory to "packages" directory

The manifest folders lived in a folder named after one in the Ansible project. This didn't make sense here, where it often lives inside a /files/ directory already.

This has been changed.

ryanrichter94 commented on Set-SslSecurity.ps1:
```diff
@@ -201,22 +215,34 @@ process {
     Update-JsonFile -Path "$env:SystemDrive\choco-setup\logs\nexus.json" -Properties @{
         NexusUri = "https://$($SubjectWithoutCn):8443"
-        NexusRepo = $RepositoryUrl
+        NexusRepo = $CoreRepositoryUrl
```
Turns out we aren't using it anyway

ryanrichter94 commented on scripts/ClientSetup.ps1:
```diff
@@ -86,13 +86,13 @@ if ($Credential) {
     $QueryString = "((Id eq 'chocolatey') and (not IsPrerelease)) and IsLatestVersion"
     $Query = 'Packages()?$filter={0}' -f [uri]::EscapeUriString($queryString)
-    $QueryUrl = ($RepositoryUrl.TrimEnd('/index.json'), $Query) -join '/'
+    $QueryUrl = ($RepositoryUrl.Replace('/index.json',''), $Query) -join '/'
```
Good to ignore. Will always be at end of URL.

ryanrichter94 commented on scripts/ClientSetup.ps1:
```diff
@@ -86,13 +86,13 @@ if ($Credential) {
 } else {
     # Otherwise, assume the URL
-    "$($RepositoryUrl.TrimEnd('/index.json'))/chocolatey/$($ChocolateyVersion)"
+    "$($RepositoryUrl.Replace('/index.json',''))/chocolatey/$($ChocolateyVersion)"
```
Good to ignore. Will always be at end of URL.

ryanrichter94 commented on packages/jenkins.json:
```diff
@@ -15,19 +15,19 @@
+        { "name": "workflow-durable-task-step", "version": "1360.v82d13453da_a_f" },
         { "name": "workflow-job", "version": "1400.v7fd111b_ec82f" },
-        { "name": "workflow-multibranch", "version": "783.va_6eb_ef636fb_d" },
+        { "name": "workflow-multibranch", "version": "791.v28fb_f74dfca_e" },
```
Not merging this with this PR

steviecoaster commented on jenkins/Update ChocolateyCore Repository/builds/permalinks:
Why is this file here?

ryanrichter94 replied:
Did some testing and all that is needed is the config.xml for the jobs. Jenkins auto-creates the rest when the build is called the first time. Removed unneeded files.

steviecoaster commented on jenkins/Update ChocolateyCore Repository/builds/legacyIds:
Why is this file here?

steviecoaster commented on jenkins/Update ChocolateyCore Repository/nextBuildNumber:
Why is this file here?

ryanrichter94 replied:
This is how Jenkins iterates build number for the jobs.

steviecoaster replied:
I don't think it has anything to do with setting up a job in Jenkins and needs removed.

steviecoaster commented on scripts/ClientSetup.ps1:
```diff
@@ -112,21 +112,28 @@ choco config set cacheLocation $env:ChocolateyInstall\choco-cache
 
 if ($InternetEnabled) {
-    choco source add --name="'ChocolateyInternal'" --source="'$RepositoryUrl'" --allow-self-service --user="'$($Credential.UserName)'" --password="'$($Credential.GetNetworkCredential().Password)'" --priority=1
+    choco source add --name="'ChocolateyCore'" --source="'$RepositoryUrl'" --allow-self-service --admin-only --user="'$($Credential.UserName)'" --password="'$($Credential.GetNetworkCredential().Password)'" --priority=0
```
Think it a good idea to have the admin-only source configured with the same credentials as the non-admin source? What if the non-admin creds get found out? Now you got 2 problems, not 1.

This requires bigger changes

ryanrichter94 replied:
The only reason ChocolateyCore repo has admin-only is to hide it from the end user GUI view. Also the credential is the same account with permissions to browse and download out of a raw/NuGet repo that ChocolateyInternal originally had before this change.

As this script is meant for a non-admin client registration. I wouldn't feel comfortable putting a true "admin" credential on either of these sources as default behavior.

steviecoaster replied:
Admin in name only. The credential _should_ and _would_ (at least this is what I expect) to be limited to browse/read operations on this new repository. The existing credential would not have access to this repository. This is a true separation such that if _either_ credential is leaked somehow, only 1 repository is affected, and even then only in that they can _download_ the packages, they'll still not be able to delete or write anything.

steviecoaster reviewed (changes requested):
_Sorry_

JPRuskin commented on Start-C4bNexusSetup.ps1:
```diff
@@ -65,12 +66,12 @@ process {
 
     # Temporary workaround to reset the NuGet v3 cache, such that it doesn't capture localhost as the FQDN
-    Remove-NexusRepositoryFolder -RepositoryName ChocolateyInternal -Name v3
+    Remove-NexusRepositoryFolder -RepositoryName ChocolateyCore -Name v3
```
FWIW, as I understand this, we're no longer pushing anything to ChocolateyInternal so there's no v3 stuff to clear.

ryanrichter94 commented on Start-C4bNexusSetup.ps1:
```diff
@@ -91,21 +92,28 @@ process {
-    if (-not (Test-Path 'C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe')) {
+    if (Test-Path 'C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe'){
+        Write-Host "Syncing Microsoft Edge, to bring it under Chocolatey management"
+        choco sync --id="Microsoft Edge" --package-id="microsoft-edge"
```
Server 2019 does not ship with Microsoft Edge

steviecoaster replied:
Yes, we should. I'm just warning that in cases where Edge has been installed via servicing channel (think a Windows Update), choco sync does _not_ detect it, even though it's installed, so we might drop out of the check erroneously.

It might be simpler to just go "screw it, I'm putting edge on here", install the package, and let the underlying install sort things out.

I bring this up because I've _seen_ sync not work even though Edge was _very clearly_ displaying in Programs and Features. It's a real peach.

ryanrichter94 commented on README.md:
```diff
@@ -34,14 +34,14 @@ As illustrated in the diagram above, there are four main components to a Chocola
     - Installation of the Chocolatey Licensed extension (`chocolatey.extension`), giving you access to features like Package Builder, Package Internalizer, etc. (full list [here](https://docs.chocolatey.org/en-us/features/)).
 
-1. **NuGet V3 Repository Server App (Nexus)**: Chocolatey works best with a NuGet V3 repository. This application hosts and manages versioning of your Chocolatey package artifacts, in their enhanced NuGet package (.nupkg) file format. The quick start guide helps you setup [Sonatype Nexus Repository Manager (OSS)](https://www.sonatype.com/products/nexus-repository).
+1. **NuGet V3 Repository Server App (Nexus)**: Chocolatey works best with a NuGet V3 repository. This application hosts and manages the versioning of your Chocolatey package artifacts, in their enhanced NuGet package (.nupkg) file format. The quick start guide helps you set up [Sonatype Nexus Repository Manager (OSS)](https://www.sonatype.com/products/nexus-repository).
```
Fixed up

ryanrichter94 commented on README.md:
```diff
@@ -34,14 +34,14 @@ As illustrated in the diagram above, there are four main components to a Chocola
+1. **NuGet V3 Repository Server App (Nexus)**: Chocolatey works best with a NuGet V3 repository. This application hosts and manages the versioning of your Chocolatey package artifacts, in their enhanced NuGet package (.nupkg) file format. The quick start guide helps you set up [Sonatype Nexus Repository Manager (OSS)](https://www.sonatype.com/products/nexus-repository).
 
-1. **Chocolatey Central Management (CCM)**: CCM is the Web UI portal for your entire Chocolatey environment. Your endpoints check-in to CCM to report their package status. This includes the Chocolatey packages they have installed, and whether any of these packages are outdated. And now, with CCM Deployments, you can also deploy packages or package updates to groups of endpoints, as well as ad-hoc ...
+1. **Chocolatey Central Management (CCM)**: CCM is the Web UI portal for your entire Chocolatey environment. Your endpoints check in to CCM to report their package status. This includes the Chocolatey packages they have installed, and whether any of these packages are outdated. And now, with CCM Deployments, you can also deploy packages or package updates to groups of endpoints, as well as ad-hoc ...
```
Added

ryanrichter94 commented:
I'm going to close this PR as the guide has changed a ton since it was last worked on. I will create a new PR for this issue.
